src/engine/authentication.md

```diff
@@ -35,10 +35,24 @@ The HMAC algorithm implies that several CLs will be able to use the same key, an
 
 The `EL` and `CL` clients **MUST** accept a cli/config parameter: `jwt-secret`, which designates a file containing the hex-encoded 256 bit secret key to be used for verifying/generating JWT tokens.
 
-If such a parameter is not given, the client **SHOULD** generate such a token, valid for the duration of the execution, and store the hex-encoded secret as a `jwt.hex` file on the filesystem.  This file can then be used to provision the counterpart client.
-
 If such a parameter _is_ given, but the file cannot be read, or does not contain a hex-encoded key of `256` bits, the client should treat this as an error: either abort the startup, or show error and continue without exposing the authenticated port.
 
+If such a parameter is not given, the client **MUST** attempt to read the secret from the default paths defined below, in the order they are listed.  If a secret is found, but the file cannot be read, does not contain a hex-encoded key of `256` bit, or is rejected by the other client, the client should continue searching. The CL **MAY** search other locations, such as default EL data directories.
+
+If no existing JWT is located, the client **MUST** generate such a token, valid for the duration of the execution, and store the hex-encoded secret on the filesystem.  The exact path is dependant on the host OS, see the default paths below.  If the client is unable to write the secret to an OS-specific path, it **MUST** fallback to writing to its own data directory.
```

Comment:
I'm not a fan of requiring tokens be stored on disk.  A high security client build may do key management through an actually secure key management service and writing the key to disk would negate those security benefits.  Similarly, a high security client may entirely remove the option of automatically generating the token in order to protect against mis-configuration leading to a security breach and such a client shouldn't be considered out of spec.

Note: Reading over this document I see some other areas where the wording is too strong that I'll try to address in a different PR.

Reply:
I think the intent of the JWT secret might play a role in the comment you made. The JWT secret is not meant to encrypt, it it meant to authenticate that the user meant to communicate with the given port. The vector that is being in case a user mistakenly exposes their engine API port to the outside world, that their EL will not be led astray. The authentication is purely meant to reduce the user error space, not to provide a secure medium of communication. 

In a scenario in which anyone has access to the disk and can read the JWT secret, we can assume the machine is completely compromised and protecting the user from an exposed port is not as high of a worry. Once you have disk level access, it would be relatively trivial to mess with the users EL, e.g: modifying their chain database or importing a new one.

Reply:
> In a scenario in which anyone has access to the disk and can read the JWT secret, we can assume the machine is completely compromised and protecting the user from an exposed port is not as high of a worry. Once you have disk level access, it would be relatively trivial to mess with the users EL, e.g: modifying their chain database or importing a new one.

Situations where a host, network, or deployment may be partially compromised and writing the JWT to disk provides an escalation vector to the attacker:
* You may have read access to the device but not write access.
* You may have disk access to the location the JWTs are stored, but not to the location the application or DB is stored (especially likely if JWTs are written to a different directory structure as proposed here).
* You have access to a backup (historical read access) but do not have access to the live machine (no write access, no realtime read access).

In general, following a security in depth strategy should not cause someone to go out of spec compliance, and not writing secrets to disk is one layer of many that helps mitigate the chances of critical failure.

Reply:
I think this is mostly out-of-scope for this PR. The spec already requires the JWT token be written to a location both clients have access to. Almost certainly, this will be a relatively permissive location for other users to read. It is of course possible to manually configure that the keys are written to a more secure location. I think general security considerations with the JWT approach should be taken up in a separate PR.

Reply:
> The spec already requires the JWT token be written to a location both clients have access to. 

Hmm, you are right.  I'll submit a PR to fix that as well.